Restart button causes error

When you expand or shrink the interface using the click links in upper left corner the next screen shows two buttons Restart and Settings.  Settings seems to work ok but exit from there returns to the Restart and Settings buttons.  If you click on restart button the following error happens and the interface is basically locked at Restart and Settings button, the only way to get out and start a game is to expand or shrink back to the starting size.

C:\Users\richa_f2oanxh\.jdks\openjdk-23.0.2\bin\java.exe -Dfile.encoding=UTF-8 -Dsun.stdout.encoding=UTF-8 -Dsun.stderr.encoding=UTF-8 -jar D:\work\source\java\Rotp-Fusion\target\rotp-Fusion-2025-08-13-mini.jar
maxMB:40707  freeMB:25420  allocMb:10184   bits:64
D:\work\source\java\Rotp-Fusion\targetRemnants.cfg not found.
@ IDE detected
@ ==================== Create newGameOptions (Setup) ==========
@ ==================== Create GameSession =====================
@ ==================== Create newGameOptions (Setup) ==========
D:\work\source\java\Rotp-Fusion\targetRemnants.cfg not found.
File not found: D:\work\source\java\Rotp-Fusion\target\Live.options
@ ==================== reset all options() ====================
@ ==================== loadStartupOptions() ===================
OS = WINDOWS
AbstractOptionsUI SingleInit()
Error attempting restart: 
java.io.IOException: Cannot run program "rotp-Fusion-2025-08-13.exe": CreateProcess error=2, The system cannot find the file specified
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1170)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1089)
	at java.base/java.lang.Runtime.exec(Runtime.java:681)
	at java.base/java.lang.Runtime.exec(Runtime.java:491)
	at java.base/java.lang.Runtime.exec(Runtime.java:366)
	at rotp.Rotp.restart(Rotp.java:326)
	at rotp.ui.game.GameUI.restartGame(GameUI.java:1196)
	at rotp.ui.game.GameUI.mouseReleased(GameUI.java:1263)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6621)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3404)
	at java.desktop/java.awt.Component.processEvent(Component.java:6386)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4996)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4828)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4948)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4575)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4516)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2780)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4828)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:775)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:98)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:747)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:87)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:744)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
	at java.base/java.lang.ProcessImpl.create(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:506)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:159)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1126)
	... 37 more

--- a/src/rotp/Rotp.java
+++ b/src/rotp/Rotp.java
@@ -72,7 +72,7 @@ public final class Rotp {
     public static String jarFileName = "rotp-" + version + RotpGovernor.miniSuffix() + ".jar";
     public static String exeFileName = "rotp-" + version + ".exe";
     public static boolean countWords = false;
-    private static String startupDir;
+    static String startupDir;
     private static boolean underTest = false; // TO DO BR: set to false
     private static Boolean isIDE;
     private static JFrame frame;
@@ -318,18 +318,35 @@ public final class Rotp {
             System.exit(0);
         }
     }
-	public static void restart() {
-        File exeFile = new File(startupDir+"/"+exeFileName);
-        String execStr = exeFile.exists() ? exeFileName : actualAlloc < 0 ? "java -jar "+jarFileName : "java -Xmx"+actualAlloc+"m -jar "+jarFileName+" arg1";
 
+    public static void restart() {
+        String dir = (startupDir != null) ? startupDir : jarPath();
+        File workingDir = new File(dir);
+        String[] command = getCommandStringTokens(workingDir);
         try {
-            Runtime.getRuntime().exec(execStr);
+            Runtime.getRuntime().exec(command, null, workingDir);
             System.exit(0);
-        } catch (IOException ex) {
+        } catch (java.io.IOException ex) {
             System.err.println("Error attempting restart: ");
             ex.printStackTrace();
         }
     }
+
+    static String[] getCommandStringTokens(File workingDir) {
+        // building tokens of a command line was a good idea to break into a function for clarity and ease of change
+        File exeFile = new File(workingDir, exeFileName);
+        String[] command;
+        // sometimes when decision logic is complex, the use of if-else is clearer than a single line conditional operator, or if word wrap and character count is long but not so complex of logic
+        if (exeFile.exists()) {
+            command = new String[] { exeFile.getAbsolutePath() };
+        } else if (actualAlloc < 0) {
+            command = new String[] { "java", "-jar", jarFileName };
+        } else {
+            command = new String[] { "java", "-Xmx" + actualAlloc + "m", "-jar", jarFileName, "arg1" };
+        }
+        return command;
+    }
+
     public static void restartFromLowMemory() {
         restartWithMoreMemory(frame, true);
     }

--- a/src/rotp/RotpGovernor.java
+++ b/src/rotp/RotpGovernor.java
@@ -165,7 +165,7 @@ public class RotpGovernor {
     private static String readBuild(String key) throws IOException {
         try (InputStream is = RotpGovernor.class.getResourceAsStream("/build.properties")) {
             if (is == null) {
-                throw new FileNotFoundException("Unable to find rotp-version.properties");
+                throw new FileNotFoundException("Unable to find build.properties");
             }
             Properties properties = new Properties();
             properties.load(is);

--- a/test/resources/build.properties
+++ b/test/resources/build.properties
@@ -1,0 +1,2 @@
+version=${project.version}
+modId=Fusion

--- a/test/rotp/RotpTest.java
+++ b/test/rotp/RotpTest.java
@@ -1,0 +1,122 @@
+package rotp;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.MockedStatic;
+
+import java.io.*;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+class RotpTest {
+
+    @BeforeEach
+    void resetStaticFields() {
+        Rotp.actualAlloc = -1;
+        Rotp.exeFileName = "test.exe";
+        Rotp.jarFileName = "test.jar";
+    }
+
+    @Test
+    void restart_UsesExeFile_IfExists() throws IOException {
+        File tempDir = new File(System.getProperty("java.io.tmpdir"));
+        File exeFile = new File(tempDir, "test.exe");
+        if (!exeFile.exists()) {
+            assertTrue(exeFile.createNewFile());
+        }
+        exeFile.deleteOnExit();
+
+        Rotp.startupDir = tempDir.getAbsolutePath();
+
+        String[] command = Rotp.getCommandStringTokens(tempDir);
+        assertEquals(exeFile.getAbsolutePath(), command[0]);
+    }
+
+    @Test
+    void restart_ExeFileCommand_ShouldIncludeExeFilePath() throws Exception {
+        File tempDir = new File(System.getProperty("java.io.tmpdir"));
+        String tempDirPath = tempDir.getAbsolutePath() + File.separator;
+        File exeFile = new File(tempDir, "test.exe");
+        if (!exeFile.exists()) {
+            assertTrue(exeFile.createNewFile());
+        }
+        exeFile.deleteOnExit();
+
+        Rotp.startupDir = tempDir.getAbsolutePath();
+        Rotp.exeFileName = "test.exe";
+        Rotp.jarFileName = "test.jar";
+        Rotp.actualAlloc = -1;
+
+        // Mock Runtime to intercept exec call
+        Runtime mockRuntime = mock(Runtime.class);
+        try (MockedStatic<Runtime> runtimeMock = mockStatic(Runtime.class, CALLS_REAL_METHODS)) {
+            runtimeMock.when(Runtime::getRuntime).thenReturn(mockRuntime);
+
+            Rotp.restart();
+            verify(mockRuntime).exec(
+                    argThat((String[] cmd) -> cmd.length > 0 && exeFile.getAbsolutePath().equals(cmd[0])), any(),
+                    any());
+        }
+    }
+
+    @Test
+    void restart_FallsBackToJar_IfExeDoesNotExist() {
+        File tempDir = new File(System.getProperty("java.io.tmpdir"));
+        File exeFile = new File(tempDir, "test.exe");
+        if (exeFile.exists()) {
+            assertTrue(exeFile.delete());
+        }
+
+        Rotp.startupDir = tempDir.getAbsolutePath();
+
+        String[] command = Rotp.getCommandStringTokens(tempDir);
+        assertEquals("java", command[0]);
+        assertEquals("-jar", command[1]);
+        assertEquals("test.jar", command[2]);
+    }
+
+    @Test
+    void restart_UsesMemoryAllocation_IfSpecified() {
+        Rotp.actualAlloc = 512;
+        File tempDir = new File(System.getProperty("java.io.tmpdir"));
+        File exeFile = new File(tempDir, "test.exe");
+        if (exeFile.exists()) {
+            assertTrue(exeFile.delete());
+        }
+
+        Rotp.startupDir = tempDir.getAbsolutePath();
+
+        String[] command = Rotp.getCommandStringTokens(tempDir);
+        assertEquals("java", command[0]);
+        assertEquals("-Xmx512m", command[1]);
+        assertEquals("-jar", command[2]);
+        assertEquals("test.jar", command[3]);
+    }
+
+    @Test
+    void restart_HandlesIOException_ThenPrintsStacktrace() throws Exception {
+        // Redirect System.err to capture output
+        PrintStream originalErr = System.err;
+        ByteArrayOutputStream errContent = new ByteArrayOutputStream();
+        System.setErr(new PrintStream(errContent));
+        // Mock Runtime.getRuntime().exec to throw IOException
+        Runtime mockRuntime = mock(Runtime.class);
+        when(mockRuntime.exec(any(String[].class), isNull(), any(File.class))).thenThrow(
+                new IOException("Simulated IO error"));
+        try (MockedStatic<Runtime> runtimeMock = mockStatic(Runtime.class, CALLS_REAL_METHODS)) {
+            runtimeMock.when(Runtime::getRuntime).thenReturn(mockRuntime);
+            // Call the method under test
+            Rotp.restart();
+        } finally {
+            System.setErr(originalErr);
+        }
+        // Assert that the error message and stack trace were printed
+        String errOutput = errContent.toString();
+        assertTrue(errOutput.contains("Error attempting restart"));
+        assertTrue(errOutput.contains("Simulated IO error"));
+        assertTrue(errOutput.contains("java.io.IOException"));
+    }
+}

